### PR TITLE
[PT2] MinMax

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

Introduce TORCH2 backend
MinMax algorithms for torch2 backend
Add handle_torch_function for quantization function to trace it by torch_function

### Related tickets

152996

### Tests
[test install](https://github.com/openvinotoolkit/nncf/actions/runs/13014595451)

### CODE REVIEW

File: tests/torch2/utils.py [hunk 1]
```diff
@@ -37,3 +41,22 @@ def compare_with_reference_file(text_data: str, ref_path: Path, regen_ref_data:
         f"Data mismatch between actual data and reference file: {ref_path}\n"
         f"Actual data and reference data differ. Please review the file contents."
     )
+
+
+def get_reference_graph(graph: NNCFGraph) -> nx.DiGraph:
```
Comment (daniil-lyakhov) [inline tests/torch2/utils.py:46]: Is it possible to reuse the `nncf_graph.get_graph_for_structure_analysis` method here?

**Review by daniil-lyakhov — COMMENTED** [review]
Great job!
It would be great if you would separate the small typo fixes from the min-max implementation, it's too much small changes for me (and probably for others as well)

**Comment by daniil-lyakhov** [comment]
Are conformance tests numbers available?

**Comment by AlexanderDokuchaev** [comment]
Model | metric | FQ | int8 | metric (N) | FQ (N) | int8   (N)
-- | -- | -- | -- | -- | -- | --
torchvision/resnet18 | 0.6945 | 30 | 21 | 0.69488 | 30 | 21
timm/mobilenetv3_small_050 | 0.4261 | 62 | 36 | 0.4192 | 62 | 36
timm/deit3_small_patch16_224 | 0.8126 | 74 | 50 | 0.8126 | 74 | 50
timm/crossvit_9_240 | 0.7275 | 112 | 88 | 0.72746 | 112 | 88
hf/bert-base-uncased |   | 74 | 77 |   | 74 | 77


(N) - experimental tracing without F/BC

File: nncf/experimental/torch2/engine.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+from typing import Any, Dict, Tuple, Union
+
+import torch
+
+from nncf.common.engine import Engine
+from nncf.experimental.torch2.function_hook.nncf_graph.nncf_graph_builder import GraphModelWrapper
+
+
+class PT2Engine(Engine):
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/engine.py:20]: We can , but what a reason? 
I have try to keep one way dependence torch2 from torch

File: nncf/experimental/torch2/model_transformer.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        model = self._model.model
+
+        for transformation_cls, transformation_fn in self._command_transformation_ordered_pairs:
+            transformations = aggregated_transformations[transformation_cls]
+            if transformations:
+                model = transformation_fn(model, transformations)  # type: ignore[arg-type]
+        return self._model
+
+    def _apply_insertion_transformation(
+        self, model: nn.Module, transformations: List[PT2InsertionCommand]
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/model_transformer.py:66]: don't see any advantages in this approach (using function instead of commands), but it makes development and debugging much more difficult.
This discuss should be part of refactoring of ModelTransformer,

File: tests/torch2/utils.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply tests/torch2/utils.py:46]: No, we have already talked about this.

File: nncf/experimental/torch2/engine.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch2/engine.py:20]: I was asked to remove the `PTFXEngine` by @alexsu52, await the same comment from him

File: nncf/experimental/torch2/model_transformer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch2/model_transformer.py:66]: Ok, I have to ask so be sure PT2 and FX backends are aligned. I'll approve it, but what's the difficulties with transformations? Why closures worse for the debug?

File: nncf/quantization/algorithms/min_max/torch_backend.py [hunk 4]
```diff
@@ -193,10 +199,10 @@ def _get_input_scale_shape(
         """
 
         is_weights = target_point.is_weight_target_point()
+        input_shape = nncf_graph.get_input_shape_for_insertion_point(target_point)
```
Comment (alexsu52) [inline nncf/quantization/algorithms/min_max/torch_backend.py:202]: As far as understand, there are input shapes in edges of nncf graph. However, constant nodes are not added to the nncf graph for PyTorch. Could you explain how it work for weights?

please, also remove it from the else block: 
```
        else:
            input_shape = nncf_graph.get_input_shape_for_insertion_point(target_point)
            channel_axes = (1,)  # channel dim for activations
```

File: nncf/quantization/algorithms/min_max/torch_backend.py [hunk 5]
```diff
@@ -167,6 +171,8 @@ def get_weight_tensor_port_ids(node: NNCFNode, graph: NNCFGraph) -> List[Optiona
 
     @staticmethod
     def get_weight_name(nncf_graph: NNCFGraph, target_point: PTTargetPoint) -> str:
+        if is_experimental_torch_tracing_enabled():
+            return target_point.target_node_name
```
Comment (alexsu52) [inline nncf/quantization/algorithms/min_max/torch_backend.py:175]: I would suggest setting layer_name to be the same as target_node_name in graph nodes to avoid adding this 'if' in all algorithms.

File: nncf/quantization/algorithms/min_max/torch_backend.py [hunk 6]
```diff
@@ -140,11 +140,15 @@ def get_start_nodes_for_activation_path_tracing(nncf_graph: PTNNCFGraph) -> List
...
-            port_id = None
-        if target_type in PTMinMaxAlgoBackend.TARGET_TYPE_TO_PT_INS_TYPE_MAP:
+            input_port_id = None
+        if (
+            not is_experimental_torch_tracing_enabled()
+            and target_type in PTMinMaxAlgoBackend.TARGET_TYPE_TO_PT_INS_TYPE_MAP
+        ):
             target_type = PTMinMaxAlgoBackend.TARGET_TYPE_TO_PT_INS_TYPE_MAP[target_type]
-        return PTTargetPoint(target_type, target_node_name, input_port_id=port_id)
+        return PTTargetPoint(target_type, target_node_name, input_port_id=input_port_id)
```
Comment (alexsu52) [inline nncf/quantization/algorithms/min_max/torch_backend.py:151]: What do you think about the same interpretation of the following target types:
- `TargetType.PRE_LAYER_OPERATION and TargetType.OPERATOR_PRE_HOOK`
- `TargetType.POST_LAYER_OPERATION: TargetType.OPERATOR_POST_HOOK`

I believe this is simplify integration of the torch2function tracing.

File: nncf/torch/quantization/layers.py [hunk 7]
```diff
@@ -290,6 +291,9 @@ def add_quantization_point(self, qp_id: QuantizationPointId, qp: PTQuantizationP
 
 
 class BaseQuantizer(nn.Module, StatefullModuleInterface, ABC):
+
+    eps: float
```
Comment (alexsu52) [inline nncf/torch/quantization/layers.py:295]: Could you explain the motivation for this change? Why does the current implementation is wrong?

File: nncf/torch/quantization/layers.py [hunk 8]
```diff
@@ -841,6 +850,8 @@ class AsymmetricQuantizer(BaseQuantizer):
     INPUT_RANGE_PARAM_NAME = "input_range"
     _INPUT_RANGE_PARAM_STORAGE_ATTR = "_input_range_param_storage"
 
+    input_low: CompressionParameter
```
Comment (alexsu52) [inline nncf/torch/quantization/layers.py:853]: Could you explain the motivation for this change? Why does the current implementation is wrong?

File: nncf/torch/quantization/layers.py [hunk 9]
```diff
@@ -1032,7 +1043,9 @@ def get_per_channel_scale_shape(input_shape, is_weights, channel_idx: int = None
     return scale_shape
 
 
-def get_scale_shape(input_shape: List[int], is_weights: bool, per_channel: bool, channel_idx: int = None) -> List[int]:
+def get_scale_shape(
+    input_shape: Sequence[int], is_weights: bool, per_channel: bool, channel_idx: int = None
```
Comment (alexsu52) [inline nncf/torch/quantization/layers.py:1047]:
```suggestion
    input_shape: Sequence[int], is_weights: bool, per_channel: bool, channel_idx: Optional[int] = None
```

File: nncf/experimental/torch2/commands.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    """
+    Insertion operation to the models.
+    """
+
+    def __init__(
+        self,
+        target_points: List[PTTargetPoint],
+        hook_module: nn.Module,
+        *,
+        handle_storage: Optional[List[RemovableHookHandle]] = None,
```
Comment (alexsu52) [inline nncf/experimental/torch2/commands.py:32]: I am confused that the developer must provide a `handle_storage` when creating the insert command. Can you explain why this is so?

File: nncf/experimental/torch2/engine.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/experimental/torch2/engine.py:20]: The reason is that you shouldn't copy and paste.

**Review by alexsu52 — COMMENTED** [review]
Do you have any performance numbers (speed and memory consumption) for t2f tracing approach?  I believe that the value of the PR will be higher if you add it to the PR description.

File: nncf/quantization/quantize_model.py [hunk 11]
```diff
@@ -58,8 +58,8 @@
 def warning_model_no_batchwise_support(
     graph: NNCFGraph,
     advanced_quantization_parameters: Optional[AdvancedQuantizationParameters],
-    model_type: ModelType,
-    no_batchwise_support_metatypes: List[OperatorMetatype],
+    model_type: Optional[ModelType],
+    no_batchwise_support_metatypes: Sequence[Type[OperatorMetatype]],
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/quantization/quantize_model.py]: Replaced to Iterable

File: nncf/torch/quantization/layers.py (hunk 7 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/quantization/layers.py]: After remove torch2 backend it no need, removed

File: nncf/torch/quantization/layers.py (hunk 8 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/quantization/layers.py]: After remove torch2 backend it no need, removed

File: nncf/torch/quantization/layers.py (hunk 9 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/quantization/layers.py]: done

File: nncf/experimental/torch2/commands.py (hunk 10 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/commands.py:32]: Do you have alternative how to return handles from ModelTransformer?

File: nncf/quantization/algorithms/min_max/torch_backend.py (hunk 6 above)
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/min_max/torch_backend.py:151]: Think that it can not simplify integration

**Comment by AlexanderDokuchaev** [comment]
Model | MinMax. time (develop) | ←PR | Stat. col. time (develop) | ←PR | RAM MiB (develop) | ←PR
:-- | --: | --: | --: | --: | --: | --:
bert-base-uncased | 0:00:56 | 0:00:52 | 0:00:53 | 0:00:49 | 535 | 553
resnet18 | 0:00:14 | 0:00:14 | 0:00:12 | 0:00:13 | 1297 | 1359
crossvit_9_240 | 0:01:03 | 0:00:56 | 0:00:59 | 0:00:53 | 1477 | 1579
deit3_small_patch16_224 | 0:00:40 | 0:00:38 | 0:00:37 | 0:00:36 | 1746 | 1854
mobilenetv3_small_050 | 0:00:24 | 0:00:24 | 0:00:22 | 0:00:22 | 1231 | 1221

File: nncf/quantization/quantize_model.py [hunk 12]
```diff
@@ -230,8 +230,10 @@ def quantize(
         )
 
     if backend == BackendType.TORCH:
-        from nncf.torch.quantization.quantize_model import quantize_impl
-
+        if os.getenv("NNCF_EXPERIMENTAL_TORCH_TRACING") is None:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/quantization/quantize_model.py]: yeap, done

File: nncf/common/factory.py [hunk 13]
```diff
@@ -129,7 +142,7 @@ def create(model: TModel) -> Engine:
 
             return PTEngine(cast(Module, model))
         raise nncf.UnsupportedBackendError(
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/common/factory.py:147]: done

**Review by alexsu52 — APPROVED** [review]
LGTM

File: nncf/experimental/torch2/commands.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    ):
+        super().__init__(TransformationType.INSERT)
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/commands.py:34]: A description of the parameters would be helpful

File: nncf/experimental/torch2/model_transformer.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
+# Copyright (c) 2024 Intel Corporation
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/model_transformer.py:1]:
```suggestion
# Copyright (c) 2024 Intel Corporation
```
```suggestion
# Copyright (c) 2025 Intel Corporation
```
Please check the copyright year

File: nncf/experimental/torch2/quantization/quantize_model.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
+# Copyright (c) 2024 Intel Corporation
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/quantization/quantize_model.py:1]:
```suggestion
# Copyright (c) 2025 Intel Corporation
```

File: nncf/quantization/algorithms/min_max/torch_backend.py [hunk 17]
```diff
@@ -140,11 +140,15 @@ def get_start_nodes_for_activation_path_tracing(nncf_graph: PTNNCFGraph) -> List
 
     @staticmethod
     def target_point(target_type: TargetType, target_node_name: str, port_id: int) -> PTTargetPoint:
+        input_port_id: Optional[int] = port_id
```
Comment (daniil-lyakhov) [inline nncf/quantization/algorithms/min_max/torch_backend.py:143]: Why input_port_id is optional here and required as the method param

File: tests/torch2/function_hook/quantization/test_quantized_graphs.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    q_model = nncf.quantize(model, nncf.Dataset([example_input]), **quantization_parameters)
+
+    if isinstance(example_input, dict):
+        nncf_graph = build_nncf_graph(q_model, **example_input)
+    else:
+        nncf_graph = build_nncf_graph(q_model, example_input)
+    nx_graph = get_reference_graph(nncf_graph)
+    dot_nncf_graph = to_pydot(nx_graph)
+    ref_file = REF_DIR / f"{desc.model_name}.dot"
+    compare_with_reference_file(str(dot_nncf_graph), ref_file, regen_ref_data)
```
Comment (daniil-lyakhov) [inline tests/torch2/function_hook/quantization/test_quantized_graphs.py:95]: What is the difference between torch and torch2 references? Is it possible to use present references?

**Review by daniil-lyakhov — COMMENTED** [review]
Minor

File: nncf/quantization/algorithms/min_max/torch_backend.py (hunk 17 above)
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/min_max/torch_backend.py:143]: Because input_port_id can be None for post operation hooks

Comment (daniil-lyakhov) [reply nncf/quantization/algorithms/min_max/torch_backend.py:143]: Please update the type hint for the port_id method param then

File: tests/torch2/function_hook/quantization/test_quantized_graphs.py (hunk 18 above)
Comment (AlexanderDokuchaev) [reply tests/torch2/function_hook/quantization/test_quantized_graphs.py:95]: Names of nodes

File: nncf/experimental/torch2/model_transformer.py (hunk 15 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/model_transformer.py]: done

File: nncf/experimental/torch2/commands.py (hunk 14 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/commands.py]: added

File: nncf/quantization/algorithms/min_max/torch_backend.py (hunk 17 above)
Comment (daniil-lyakhov) [reply nncf/quantization/algorithms/min_max/torch_backend.py:143]: Ok, got it

**Review by daniil-lyakhov — APPROVED** [review]
